core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,420 @@
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        @SuppressWarnings("unchecked")
+        IndexGraveyard that = (IndexGraveyard) obj;
+        return Objects.equals(tombstones, that.tombstones);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tombstones);
+    }
+
+    /**
+     * Get a copy of the current index tombstones.
+     */
+    List<Tombstone> tombstones() {
+        return new ArrayList<>(tombstones);
+    }
+
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startArray();
+        for (Tombstone tombstone : tombstones) {
+            tombstone.toXContent(builder, params);
+        }
+        return builder.endArray();
+    }
+
+    public static IndexGraveyard fromXContent(final XContentParser parser) throws IOException {
+        if (parser.currentToken() != XContentParser.Token.START_ARRAY) {
+            parser.nextToken();
+        }
+        List<Tombstone> tombstones = new ArrayList<>();
+        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+            tombstones.add(Tombstone.fromXContent(parser));
+        }
+        return new IndexGraveyard(tombstones);
+    }
+
+    @Override
+    public String toString() {
+        return "IndexGraveyard[" + tombstones + "]";
+    }
+
+    @Override
+    public void writeTo(final StreamOutput out) throws IOException {
+        out.writeVInt(tombstones.size());
+        for (Tombstone tombstone : tombstones) {
+            tombstone.writeTo(out);
+        }
+    }
+
+    @Override
+    public IndexGraveyard readFrom(final StreamInput in) throws IOException {
+        return new IndexGraveyard(in);
+    }
+
+    @Override
+    public IndexGraveyardDiff diff(final IndexGraveyard previous) {
+        return new IndexGraveyardDiff(previous, this);
+    }
+
+    @Override
+    public IndexGraveyardDiff readDiffFrom(final StreamInput in) throws IOException {
+        return new IndexGraveyardDiff(in);
+    }
+
+    public static IndexGraveyard.Builder builder() {
+        return new IndexGraveyard.Builder();
+    }
+
+    public static IndexGraveyard.Builder builder(final IndexGraveyard graveyard) {
+        return new IndexGraveyard.Builder(graveyard);
+    }
+
+    /**
+     * A class to build an IndexGraveyard.
+     */
+    final public static class Builder {
+        private final Deque<Tombstone> tombstones;
+
+        private Builder() {
+            tombstones = new ArrayDeque<>();
+        }
+
+        private Builder(IndexGraveyard that) {
+            tombstones = new ArrayDeque<>(that.tombstones());
+        }
+
+        private Builder(final Deque<Tombstone> tombstones) {
+            this.tombstones = tombstones;
+        }
+
+        /**
+         * A copy of the current tombstones in the builder.
+         */
+        public List<Tombstone> tombstones() {
```

Comment:
If we use Collection<Tombstonre> we can return an unmodifiableCollection() which doesn't copy stuff..